### PR TITLE
Remote function planning

### PR DESCRIPTION
**Pull request by rongrong**
This PR introduced the concept of "remote function" into planner and generates query plan that's aware of the remote functions. Depends on #14219. Partially addresses #14053.

Ran verifier explain tests across all dcs and all tests passed.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    @BeforeClass
+    public void setup()
+    {
+        FunctionManager functionManager = getFunctionManager();
+        functionManager.addTestFunctionNamespace("unittest", new InMemoryFunctionNamespaceManager("unittest", new SqlInvokedFunctionNamespaceManagerConfig().setSupportedFunctionLanguages("sql,java")));
+        functionManager.createFunction(FUNCTION_REMOTE_FOO_0, true);
+        functionManager.createFunction(FUNCTION_REMOTE_FOO_1, true);
+        functionManager.createFunction(FUNCTION_REMOTE_FOO_2, true);
+        functionManager.createFunction(FUNCTION_REMOTE_FOO_3, true);
```
Comment (caithagoras) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java:116]: Those exact setups have already been done in the constructor.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            RoutineCharacteristics.builder().setLanguage(JAVA).setNullCallClause(RETURNS_NULL_ON_NULL_INPUT).build(),
+            "",
+            Optional.empty());
+
+    private static final Metadata METADATA = MetadataManager.createTestMetadataManager();
+
+    public TestPlanRemoteProjections()
+    {
+        FunctionManager functionManager = METADATA.getFunctionManager();
+        functionManager.addTestFunctionNamespace("unittest", new InMemoryFunctionNamespaceManager("unittest", new SqlInvokedFunctionNamespaceManagerConfig().setSupportedFunctionLanguages("sql,java")));
```
Comment (caithagoras) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java:101]: Let's use H2 FNM, which is designed for testing purpose. We can then avoid adding test-related method `addTestFunctionNamespace` in our production code.

You can do that by first adding the factory:
```Java
static {
    METADATA.getFunctionManager().addFunctionNamespaceFactory(new H2FunctionNamespaceManagerFactory());
}
```

You'll then be able to add a function namespace by
```Java
functionManager.loadFunctionNamespaceManager(
    H2FunctionNamespaceManagerFactory.NAME,
    "unittest",
    ImmutableMap.of(<config key>, <config value>))
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 3]
```diff
@@ -107,4 +109,12 @@ static void verifyNoAggregateWindowOrGroupingFunctions(Map<NodeRef<FunctionCall>
             throw new SemanticException(CANNOT_HAVE_AGGREGATIONS_WINDOWS_OR_GROUPING, predicate, "%s cannot contain aggregations, window functions or grouping operations: %s", clause, found);
         }
     }
+
+    static void verifyNoExternalFunctions(Map<NodeRef<FunctionCall>, FunctionHandle> functionHandles, FunctionManager functionManager, Expression predicate, String clause)
+    {
+        List<FunctionCall> externalFunctions = extractExternalFunctions(functionHandles, ImmutableList.of(predicate), functionManager);
+        if (!externalFunctions.isEmpty()) {
+            throw new SemanticException(NOT_SUPPORTED, predicate, "External functions in %s is not supported: %s", clause, externalFunctions);
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:117]: nit, for better readability of the constructed error message:
```
External functions in [%s] is not supported: %s
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java [hunk 4]
```diff
@@ -67,6 +72,11 @@ private static boolean isWindowFunction(FunctionCall functionCall)
         return functionCall.getWindow().isPresent();
     }
 
+    private static Predicate<FunctionCall> isExternalFunctionPredicate(Map<NodeRef<FunctionCall>, FunctionHandle> functionHandles, FunctionManager functionManager)
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:75]: This private helper method is only used once. Just inline?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java [hunk 5]
```diff
@@ -67,6 +72,11 @@ private static boolean isWindowFunction(FunctionCall functionCall)
...
+    {
+        return ((functionCall) -> (functionManager.getFunctionMetadata(functionHandles.get(NodeRef.of(functionCall))).getImplementationType().isExternal()));
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:77]: Remove redundant parenthesis:
```
return function -> functionManager.getFunctionMetadata ...
````

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java (hunk 3 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:117]: There is no unit test coverage for this commit. Can we add some.

File: presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/AbstractSqlInvokedFunctionNamespaceManager.java [hunk 6]
```diff
@@ -138,18 +138,24 @@ public final void abort(FunctionNamespaceTransactionHandle transactionHandle)
...
-        checkArgument(transactionHandle.isPresent(), "missing transactionHandle");
         // This is the only assumption in this class that we're dealing with sql-invoked regular function.
         SqlFunctionId functionId = new SqlFunctionId(signature.getName(), signature.getArgumentTypes());
-        return transactions.get(transactionHandle.get()).getFunctionHandle(functionId);
+        if (transactionHandle.isPresent()) {
+            return transactions.get(transactionHandle.get()).getFunctionHandle(functionId);
+        }
+        FunctionCollection collection = new FunctionCollection();
+        collection.loadAndGetFunctionsTransactional(signature.getName());
+        return collection.getFunctionHandle(functionId);
```
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/AbstractSqlInvokedFunctionNamespaceManager.java:158]: Is transaction must exist we should change it to be not optional. As long as it's `Optional` it's strange to argue that it has to exist.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java (hunk 1 above)
Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java:116]: The two types of tests are setup differently. They are not using the same function manager. I don't remember the specifics. If you really care I can see what I can do about it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java (hunk 2 above)
Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java:101]: I remember I tried that but it didn't work. But I don't remember what didn't work. This is the problem of delayed review... Let me try again I guess...

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java:101]: I think the reason to not use `H2FunctionNamespaceManagerFactory` is dependencies. This test is in presto-main. `H2FunctionNamespaceManagerFactory` is in presto-tests.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java (hunk 4 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:75]: All the other predicates in this class are only used once. I'm just keeping them consistent (including the style) with the rest of the file.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:117]: > nit, for better readability of the constructed error message:
> 
> ```
> External functions in [%s] is not supported: %s
> ```

I don't think the [] is useful. It would make the error message like this:
```
External functions in [Lambda expression] is not supported:....
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/VerifyProjectionLocality.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                    node.getAssignments().getExpressions().forEach(expression -> {
+                        checkState(expression instanceof VariableReferenceExpression || expression instanceof CallExpression, format("Expect VariableReferenceExpression or CallExpression, got %s", expression.getClass().getName()));
+                        if (expression instanceof CallExpression) {
+                            checkState(expression.accept(externalCallExpressionChecker, null), format("Expect expression %s to be an external function", expression));
+                        }
+                    });
+                    break;
+                case UNKNOWN:
+                    throw new IllegalStateException("ProjectNode should have locality LOCAL or REMOTE");
+            }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/VerifyProjectionLocality.java:69]: Maybe add a `default:`?

**Comment by kaikalur** [comment]
Overall, looks good. It would be good to add a few tests that actually show a real plan string (like explain output) with external functions so the logic becomes clearer and will help future iterations on this feature.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 8]
```diff
@@ -169,6 +169,15 @@ public void loadFunctionNamespaceManager(
         }
     }
 
+    @VisibleForTesting
+    public void addTestFunctionNamespace(String catalogName, FunctionNamespaceManager functionNamespaceManager)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:173]: `addFunctionNamespace` should be good enough given it's for testing only
